File: javascript/section03/spaceComplexity.js
function boooo(n) {
    for (let i = 0; i < n; i++) { //o(1)
      console.log("booooo");
    }
  }
  
  function arrayOfHiNTimes(n) { //o(n)
    var hiArray = [];//o(1)
    for (let i = 0; i < n; i++) { //o(1)
      hiArray[i] = "hi"; //O(n)
    }
    return hiArray;
  }
  
  arrayOfHiNTimes(6);


/*
  The boo function has a space complexity of O(1) because it's allocatiion a constant amount of memoery when declaring let i in the for loop
  The arrayOfHiNTimes has a space complexity of O(n) because were adding a new element that number of the input. So each item is an additional memory space on our computers 
*/

File: cpp/section03/spaceComplexity.cpp
#include <iostream>
#include <vector>
#include <string>

// Function to print "booooo" n times
void boooo(int n) {
    for (int i = 0; i < n; i++) { // O(n)
        std::cout << "booooo" << std::endl;
    }
}

// Function to create an array of "hi" n times
std::vector<std::string> arrayOfHiNTimes(int n) { // O(n)
    std::vector<std::string> hiArray(n); // O(n) - Allocate space for n elements

    for (int i = 0; i < n; i++) { // O(n)
        hiArray[i] = "hi"; // O(1) - Assign "hi" to each element
    }

    return hiArray;
}

int main() {
    // Call the function to create an array of "hi" 6 times
    std::vector<std::string> result = arrayOfHiNTimes(6);
    
    // Print the resulting array
    for (const auto& hi : result) {
        std::cout << hi << std::endl;
    }

    return 0;
}

/*
    The boooo function has a space complexity of O(1) because it allocates a constant amount of memory
    when declaring the variable i in the for loop.

    The arrayOfHiNTimes function has a space complexity of O(n) because we are creating a vector of size n,
    and each element represents an additional memory allocation on our computer.
*/
